pkg/pipeline/extract/aggregate/aggregates.go

```diff
@@ -54,20 +63,39 @@ func (aggregates Aggregates) GetMetrics() []config.GenericMap {
 func (aggregates Aggregates) AddAggregate(aggregateDefinition api.AggregateDefinition) Aggregates {
 	aggregate := Aggregate{
 		Definition: aggregateDefinition,
-		Groups:     map[NormalizedValues]*GroupState{},
+		GroupsMap:  map[NormalizedValues]*GroupState{},
+		GroupsList: list.New(),
 	}
 
 	appendedAggregates := append(aggregates, aggregate)
 	return appendedAggregates
 }
 
-func (aggregates Aggregates) RemoveAggregate(by api.AggregateBy) (Aggregates, error) {
-	for i, other := range aggregates {
-		if reflect.DeepEqual(other.Definition.By, by) {
-			return append(aggregates[:i], aggregates[i+1:]...), nil
+func (aggregates Aggregates) cleanupExpiredEntriesLoop() {
+
+	ticker := time.NewTicker(time.Duration(expiryTime) * time.Second)
+	done := make(chan bool)
+	utils.RegisterExitChannel(done)
+	go func() {
+		for {
+			select {
+			case <-done:
+				return
+			case <-ticker.C:
+				aggregates.cleanupExpiredEntries()
+			}
 		}
+	}()
+}
+
+func (aggregates Aggregates) cleanupExpiredEntries() {
+	mutex.Lock()
+	defer mutex.Unlock()
+
+	for _, aggregate := range aggregates {
+		aggregate.cleanupExpiredEntries()
```

Comment:
Each aggregate has its own cleanup loop. If so, then can conflict on the common mutext lock. Why don't you put a separate mutex on each aggregate so there won't be any conflicts between them?

Reply:
done